assets/js/hocs/with-searched-products.js

```diff
@@ -0,0 +1,86 @@
+/**
+ * External dependencies
+ */
+import { Component } from '@wordpress/element';
+import { debounce, find } from 'lodash';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import { isLargeCatalog, getProducts } from '../components/utils';
+
+/**
+ * A higher order component that enhances the provided component with products
+ * from a search query.
+ */
+const withSearchedProducts = createHigherOrderComponent( ( OriginalComponent ) => {
+	/**
+	 * A Component wrapping the passed in component.
+	 *
+	 * @class WrappedComponent
+	 * @extends {Component}
+	 */
+	class WrappedComponent extends Component {
+		constructor() {
+			super( ...arguments );
+			this.state = {
+				list: [],
+				loading: true,
+			};
+			this.debouncedOnSearch = debounce( this.onSearch.bind( this ), 400 );
+		}
+
+		componentDidMount() {
+			const { selected } = this.props;
+			getProducts( { selected } )
+				.then( ( list ) => {
+					this.setState( { list, loading: false } );
+				} )
+				.catch( () => {
+					this.setState( { list: [], loading: false } );
+				} );
+		}
+
+		componentWillUnmount() {
+			this.debouncedOnSearch.cancel();
+		}
+
+		onSearch( search ) {
+			const { selected } = this.props;
+			getProducts( { selected, search } )
+				.then( ( list ) => {
+					this.setState( { list, loading: false } );
+				} )
+				.catch( () => {
+					this.setState( { list: [], loading: false } );
+				} );
+		}
+
+		render() {
+			const { list, loading } = this.state;
+			const { selected } = this.props;
+			return (
+				<OriginalComponent
+					{ ...this.props }
+					products={ list }
+					isLoading={ loading }
+					selected={ selected
+						.map( ( id ) => find( list, { id } ) )
+						.filter( Boolean ) }
```

Comment:
I know that's not new from this PR either, but if I'm not missing any edge case this can be rewritten and avoid the lodash dependency:
```ES
selected={ list.filter( ( { id } ) => selected.includes( id ) ) }
```

Reply:
fixed in e48b140